File: cpp/map_core/src/geometry.cpp
#include "automap/core/geometry.hpp"

#include <algorithm>
#include <cmath>
#include <numbers>
#include <type_traits>
#include <utility>

namespace automap::core {
namespace {

[[nodiscard]] double horizontal_distance(
    const Point3d& first,
    const Point3d& second) noexcept {
    return std::hypot(second.x - first.x, second.y - first.y);
}

template <typename Function>
[[nodiscard]] double adaptive_simpson(
    const Function& function,
    double start,
    double end,
    double tolerance,
    std::size_t depth) noexcept {
    const auto simpson = [&](double left, double right) {
        const double middle = (left + right) * 0.5;
        return (right - left) / 6.0 *
               (function(left) + 4.0 * function(middle) + function(right));
    };

    const double whole = simpson(start, end);
    const auto recurse = [&](auto&& self, double left, double right,
                             double expected, double epsilon,
                             std::size_t remaining) -> double {
        const double middle = (left + right) * 0.5;
        const double first = simpson(left, middle);
        const double second = simpson(middle, right);
        const double delta = first + second - expected;
        if (remaining == 0U || std::abs(delta) <= 15.0 * epsilon) {
            return first + second + delta / 15.0;
        }
        return self(self, left, middle, first, epsilon * 0.5, remaining - 1U) +
               self(self, middle, right, second, epsilon * 0.5, remaining - 1U);
    };
    return recurse(recurse, start, end, whole, tolerance, depth);
}

[[nodiscard]] CurveState3d evaluate_forward_segment(
    const CurveSegment3d& segment,
    double requested_s) noexcept {
    return std::visit(
        [requested_s](const auto& value) -> CurveState3d {
            const double length = std::max(0.0, value.length_m);
            const double s = std::clamp(requested_s, 0.0, length);
            const double z_ratio = length > 0.0 ? s / length : 0.0;
            const double z = value.start.z + (value.end_z_m - value.start.z) * z_ratio;

            using Segment = std::decay_t<decltype(value)>;
            if constexpr (std::is_same_v<Segment, LineCurveSegment3d>) {
                return {
                    .position = {
                        value.start.x + std::cos(value.heading_rad) * s,
                        value.start.y + std::sin(value.heading_rad) * s,
                        z,
                    },
                    .heading_rad = normalize_heading(value.heading_rad),
                    .curvature_per_m = 0.0,
                };
            } else if constexpr (std::is_same_v<Segment, CircularArcSegment3d>) {
                const double curvature = value.curvature_per_m;
                if (std::abs(curvature) <= tolerance::floating_point) {
                    return {
                        .position = {
                            value.start.x + std::cos(value.heading_rad) * s,
                            value.start.y + std::sin(value.heading_rad) * s,
                            z,
                        },
                        .heading_rad = normalize_heading(value.heading_rad),
                        .curvature_per_m = curvature,
                    };
                }
                const double heading = value.heading_rad + curvature * s;
                return {
                    .position = {
                        value.start.x +
                            (std::sin(heading) - std::sin(value.heading_rad)) / curvature,
                        value.start.y -
                            (std::cos(heading) - std::cos(value.heading_rad)) / curvature,
                        z,
                    },
                    .heading_rad = normalize_heading(heading),
                    .curvature_per_m = curvature,
                };
            } else {
                const double curvature_rate = length > 0.0
                    ? (value.end_curvature_per_m - value.start_curvature_per_m) / length
                    : 0.0;
                const auto heading_at = [&](double distance_m) {
                    return value.heading_rad + value.start_curvature_per_m * distance_m +
                           0.5 * curvature_rate * distance_m * distance_m;
                };
                const double integration_tolerance =
                    std::max(1e-12, std::abs(s) * 1e-12);
                const double offset_x = s == 0.0 ? 0.0 : adaptive_simpson(
                    [&](double distance_m) { return std::cos(heading_at(distance_m)); },
                    0.0, s, integration_tolerance, 18U);
                const double offset_y = s == 0.0 ? 0.0 : adaptive_simpson(
                    [&](double distance_m) { return std::sin(heading_at(distance_m)); },
                    0.0, s, integration_tolerance, 18U);
                return {
                    .position = {value.start.x + offset_x, value.start.y + offset_y, z},
                    .heading_rad = normalize_heading(heading_at(s)),
                    .curvature_per_m = value.start_curvature_per_m + curvature_rate * s,
                };
            }
        },
        segment);
}

void append_tessellated_interval(
    const CurveSegment3d& segment,
    double start_s,
    const CurveState3d& start_state,
    double end_s,
    const CurveState3d& end_state,
    const TessellationOptions& options,
    std::size_t depth,
    Polyline3d& result) {
    const double middle_s = (start_s + end_s) * 0.5;
    const CurveState3d middle_state = evaluate_curve_segment(segment, middle_s);
    const Point3d chord_middle{
        (start_state.position.x + end_state.position.x) * 0.5,
        (start_state.position.y + end_state.position.y) * 0.5,
        (start_state.position.z + end_state.position.z) * 0.5,
    };
    const double chord_error = horizontal_distance(middle_state.position, chord_middle);
    const double interval_length = end_s - start_s;
    const bool subdivide = depth < options.max_recursion_depth &&
        (chord_error > options.max_chord_error_m ||
         interval_length > options.max_segment_length_m);
    if (subdivide) {
        append_tessellated_interval(
            segment, start_s, start_state, middle_s, middle_state,
            options, depth + 1U, result);
        append_tessellated_interval(
            segment, middle_s, middle_state, end_s, end_state,
            options, depth + 1U, result);
        return;
    }
    result.push_back(end_state.position);
}

}  // namespace

PathGeometry3d::PathGeometry3d(Polyline3d polyline)
    : storage_(std::move(polyline)) {}

PathGeometry3d::PathGeometry3d(std::initializer_list<Point3d> points)
    : storage_(Polyline3d{points}) {}

PathGeometry3d::PathGeometry3d(CompositeCurve3d curve)
    : storage_(std::move(curve)) {}

bool PathGeometry3d::is_polyline() const noexcept {
    return std::holds_alternative<Polyline3d>(storage_);
}

bool PathGeometry3d::is_composite_curve() const noexcept {
    return std::holds_alternative<CompositeCurve3d>(storage_);
}

const Polyline3d* PathGeometry3d::polyline() const noexcept {
    return std::get_if<Polyline3d>(&storage_);
}

Polyline3d* PathGeometry3d::polyline() noexcept {
    return std::get_if<Polyline3d>(&storage_);
}

const CompositeCurve3d* PathGeometry3d::composite_curve() const noexcept {
    return std::get_if<CompositeCurve3d>(&storage_);
}

CompositeCurve3d* PathGeometry3d::composite_curve() noexcept {
    return std::get_if<CompositeCurve3d>(&storage_);
}

const PathGeometry3d::Storage& PathGeometry3d::storage() const noexcept {
    return storage_;
}

bool BoundingBox3d::contains(const Point3d& point, double epsilon) const noexcept {
    return point.x >= min.x - epsilon && point.x <= max.x + epsilon &&
           point.y >= min.y - epsilon && point.y <= max.y + epsilon &&
           point.z >= min.z - epsilon && point.z <= max.z + epsilon;
}

bool almost_equal(double lhs, double rhs, double epsilon) noexcept {
    if (lhs == rhs) {
        return true;
    }

    const double scale = std::max({1.0, std::abs(lhs), std::abs(rhs)});
    return std::abs(lhs - rhs) <= epsilon * scale;
}

double distance(const Point3d& lhs, const Point3d& rhs) noexcept {
    return std::hypot(lhs.x - rhs.x, lhs.y - rhs.y, lhs.z - rhs.z);
}

bool points_coincident(
    const Point3d& lhs,
    const Point3d& rhs,
    double epsilon_m) noexcept {
    return distance(lhs, rhs) <= epsilon_m;
}

double polyline_length(const Polyline3d& polyline) noexcept {
    double length = 0.0;
    for (std::size_t index = 1; index < polyline.size(); ++index) {
        length += distance(polyline[index - 1], polyline[index]);
    }
    return length;
}

double polyline_length(const PathGeometry3d& path) noexcept {
    return path_planar_length(path);
}

std::optional<BoundingBox3d> bounding_box(const Polyline3d& polyline) noexcept {
    if (polyline.empty()) {
        return std::nullopt;
    }

    BoundingBox3d bounds{.min = polyline.front(), .max = polyline.front()};
    for (const Point3d& point : polyline) {
        bounds.min.x = std::min(bounds.min.x, point.x);
        bounds.min.y = std::min(bounds.min.y, point.y);
        bounds.min.z = std::min(bounds.min.z, point.z);
        bounds.max.x = std::max(bounds.max.x, point.x);
        bounds.max.y = std::max(bounds.max.y, point.y);
        bounds.max.z = std::max(bounds.max.z, point.z);
    }
    return bounds;
}

std::optional<BoundingBox3d> bounding_box(const PathGeometry3d& path) {
    return path_bounding_box(path);
}

std::optional<BoundingBox3d> bounding_box(
    std::initializer_list<Point3d> points) noexcept {
    return bounding_box(Polyline3d{points});
}

double normalize_heading(double heading_rad) noexcept {
    return std::remainder(heading_rad, 2.0 * std::numbers::pi_v<double>);
}

double curve_segment_length(const CurveSegment3d& segment) noexcept {
    return std::visit([](const auto& value) { return value.length_m; }, segment);
}

CurveState3d evaluate_curve_segment(
    const CurveSegment3d& segment,
    double s) noexcept {
    return evaluate_forward_segment(segment, s);
}

double path_planar_length(const PathGeometry3d& path) noexcept {
    if (const Polyline3d* polyline = path.polyline()) {
        double length = 0.0;
        for (std::size_t index = 1U; index < polyline->size(); ++index) {
            length += horizontal_distance((*polyline)[index - 1U], (*polyline)[index]);
        }
        return length;
    }

    double length = 0.0;
    for (const CurveSegment3d& segment : path.composite_curve()->segments) {
        const double segment_length = curve_segment_length(segment);
        if (std::isfinite(segment_length) && segment_length > 0.0) {
            length += segment_length;
        }
    }
    return length;
}

CurveState3d evaluate_path_geometry(
    const PathGeometry3d& path,
    double requested_s,
    PathTraversal traversal) noexcept {
    const double total_length = path_planar_length(path);
    const double traversal_s = std::clamp(requested_s, 0.0, total_length);
    const double forward_s = traversal == PathTraversal::forward
        ? traversal_s
        : total_length - traversal_s;

    CurveState3d state{};
    if (const Polyline3d* polyline = path.polyline()) {
        if (polyline->empty()) {
            return state;
        }
        state.position = polyline->front();
        double accumulated = 0.0;
        for (std::size_t index = 1U; index < polyline->size(); ++index) {
            const Point3d& first = (*polyline)[index - 1U];
            const Point3d& second = (*polyline)[index];
            const double segment_length = horizontal_distance(first, second);
            if (segment_length <= tolerance::floating_point) {
                continue;
            }
            if (forward_s <= accumulated + segment_length || index + 1U == polyline->size()) {
                const double local_s = std::clamp(forward_s - accumulated, 0.0, segment_length);
                const double ratio = local_s / segment_length;
                state.position = {
                    first.x + (second.x - first.x) * ratio,
                    first.y + (second.y - first.y) * ratio,
                    first.z + (second.z - first.z) * ratio,
                };
                state.heading_rad = std::atan2(second.y - first.y, second.x - first.x);
                break;
            }
            accumulated += segment_length;
        }
    } else {
        const auto& segments = path.composite_curve()->segments;
        if (segments.empty()) {
            return state;
        }
        double accumulated = 0.0;
        state = evaluate_curve_segment(segments.front(), 0.0);
        for (std::size_t index = 0U; index < segments.size(); ++index) {
            const double segment_length = std::max(0.0, curve_segment_length(segments[index]));
            if (forward_s <= accumulated + segment_length || index + 1U == segments.size()) {
                state = evaluate_curve_segment(
                    segments[index], std::clamp(forward_s - accumulated, 0.0, segment_length));
                break;
            }
            accumulated += segment_length;
        }
    }

    if (traversal == PathTraversal::reverse) {
        state.heading_rad = normalize_heading(state.heading_rad + std::numbers::pi_v<double>);
        state.curvature_per_m = -state.curvature_per_m;
    }
    return state;
}

CurveState3d path_start_state(
    const PathGeometry3d& path,
    PathTraversal traversal) noexcept {
    return evaluate_path_geometry(path, 0.0, traversal);
}

CurveState3d path_end_state(
    const PathGeometry3d& path,
    PathTraversal traversal) noexcept {
    return evaluate_path_geometry(path, path_planar_length(path), traversal);
}

Polyline3d tessellate_path_geometry(
    const PathGeometry3d& path,
    const TessellationOptions& requested_options) {
    if (const Polyline3d* polyline = path.polyline()) {
        return *polyline;
    }

    TessellationOptions options = requested_options;
    options.max_chord_error_m = std::max(options.max_chord_error_m, 1e-9);
    options.max_segment_length_m = std::max(options.max_segment_length_m, 1e-6);

    Polyline3d result;
    const auto& segments = path.composite_curve()->segments;
    for (const CurveSegment3d& segment : segments) {
        const double length = curve_segment_length(segment);
        const CurveState3d start = evaluate_curve_segment(segment, 0.0);
        if (result.empty() || !points_coincident(result.back(), start.position)) {
            result.push_back(start.position);
        }
        if (!std::isfinite(length) || length <= 0.0) {
            continue;
        }
        const CurveState3d end = evaluate_curve_segment(segment, length);
        append_tessellated_interval(
            segment, 0.0, start, length, end, options, 0U, result);
    }
    return result;
}

std::optional<BoundingBox3d> path_bounding_box(
    const PathGeometry3d& path,
    const TessellationOptions& options) {
    return bounding_box(tessellate_path_geometry(path, options));
}

}  // namespace automap::core

File: visualizer/src/render/geometry.ts
import type { CurveSegment3d, PathGeometry3d, Point3d } from "../model";
import { isCompositeCurve } from "../model";

export interface CurveState3d {
  position: Point3d;
  headingRad: number;
  curvaturePerM: number;
}

export interface Bounds2d {
  minX: number;
  minY: number;
  maxX: number;
  maxY: number;
}

export function normalizeHeading(value: number): number {
  const full = Math.PI * 2;
  return ((value + Math.PI) % full + full) % full - Math.PI;
}

function simpson(fn: (value: number) => number, start: number, end: number): number {
  const middle = (start + end) * 0.5;
  return (end - start) / 6 * (fn(start) + 4 * fn(middle) + fn(end));
}

function adaptiveSimpson(
  fn: (value: number) => number,
  start: number,
  end: number,
  tolerance: number,
  depth = 18,
): number {
  const recurse = (left: number, right: number, expected: number, epsilon: number, remaining: number): number => {
    const middle = (left + right) * 0.5;
    const first = simpson(fn, left, middle);
    const second = simpson(fn, middle, right);
    const delta = first + second - expected;
    if (remaining === 0 || Math.abs(delta) <= 15 * epsilon) return first + second + delta / 15;
    return recurse(left, middle, first, epsilon * 0.5, remaining - 1)
      + recurse(middle, right, second, epsilon * 0.5, remaining - 1);
  };
  return recurse(start, end, simpson(fn, start, end), tolerance, depth);
}

export function evaluateSegment(segment: CurveSegment3d, requestedS: number): CurveState3d {
  const length = Math.max(0, segment.lengthM);
  const s = Math.min(Math.max(requestedS, 0), length);
  const ratio = length > 0 ? s / length : 0;
  const z = segment.start[2] + (segment.endZM - segment.start[2]) * ratio;
  if (segment.type === "line") {
    return {
      position: [segment.start[0] + Math.cos(segment.headingRad) * s, segment.start[1] + Math.sin(segment.headingRad) * s, z],
      headingRad: normalizeHeading(segment.headingRad), curvaturePerM: 0,
    };
  }
  if (segment.type === "circular_arc") {
    const curvature = segment.curvaturePerM;
    if (Math.abs(curvature) <= 1e-12) {
      return {
        position: [segment.start[0] + Math.cos(segment.headingRad) * s, segment.start[1] + Math.sin(segment.headingRad) * s, z],
        headingRad: normalizeHeading(segment.headingRad), curvaturePerM: curvature,
      };
    }
    const heading = segment.headingRad + curvature * s;
    return {
      position: [
        segment.start[0] + (Math.sin(heading) - Math.sin(segment.headingRad)) / curvature,
        segment.start[1] - (Math.cos(heading) - Math.cos(segment.headingRad)) / curvature,
        z,
      ],
      headingRad: normalizeHeading(heading), curvaturePerM: curvature,
    };
  }
  const rate = length > 0 ? (segment.endCurvaturePerM - segment.startCurvaturePerM) / length : 0;
  const headingAt = (distance: number): number => segment.headingRad
    + segment.startCurvaturePerM * distance + 0.5 * rate * distance * distance;
  const tolerance = Math.max(1e-12, Math.abs(s) * 1e-12);
  const x = s === 0 ? 0 : adaptiveSimpson((distance) => Math.cos(headingAt(distance)), 0, s, tolerance);
  const y = s === 0 ? 0 : adaptiveSimpson((distance) => Math.sin(headingAt(distance)), 0, s, tolerance);
  return {
    position: [segment.start[0] + x, segment.start[1] + y, z],
    headingRad: normalizeHeading(headingAt(s)),
    curvaturePerM: segment.startCurvaturePerM + rate * s,
  };
}

export function pathLength(geometry: PathGeometry3d): number {
  if (isCompositeCurve(geometry)) {
    return geometry.segments.reduce((total, segment) => total + Math.max(0, segment.lengthM), 0);
  }
  let total = 0;
  for (let index = 1; index < geometry.length; index += 1) {
    total += Math.hypot(geometry[index][0] - geometry[index - 1][0], geometry[index][1] - geometry[index - 1][1]);
  }
  return total;
}

export function evaluatePath(geometry: PathGeometry3d, requestedS: number): CurveState3d {
  const total = pathLength(geometry);
  const s = Math.min(Math.max(requestedS, 0), total);
  if (isCompositeCurve(geometry)) {
    let accumulated = 0;
    for (let index = 0; index < geometry.segments.length; index += 1) {
      const segment = geometry.segments[index];
      const length = Math.max(0, segment.lengthM);
      if (s <= accumulated + length || index === geometry.segments.length - 1) {
        return evaluateSegment(segment, s - accumulated);
      }
      accumulated += length;
    }
    return { position: [0, 0, 0], headingRad: 0, curvaturePerM: 0 };
  }
  if (geometry.length === 0) return { position: [0, 0, 0], headingRad: 0, curvaturePerM: 0 };
  let accumulated = 0;
  for (let index = 1; index < geometry.length; index += 1) {
    const first = geometry[index - 1];
    const second = geometry[index];
    const length = Math.hypot(second[0] - first[0], second[1] - first[1]);
    if (length <= 1e-12) continue;
    if (s <= accumulated + length || index === geometry.length - 1) {
      const ratio = Math.min(Math.max((s - accumulated) / length, 0), 1);
      return {
        position: [first[0] + (second[0] - first[0]) * ratio, first[1] + (second[1] - first[1]) * ratio, first[2] + (second[2] - first[2]) * ratio],
        headingRad: Math.atan2(second[1] - first[1], second[0] - first[0]), curvaturePerM: 0,
      };
    }
    accumulated += length;
  }
  return { position: geometry[0], headingRad: 0, curvaturePerM: 0 };
}

export function tessellatePath(
  geometry: PathGeometry3d,
  maxChordErrorM = 0.25,
  maxSegmentLengthM = 10,
): Point3d[] {
  if (!isCompositeCurve(geometry)) return geometry.map((point) => [...point] as Point3d);
  const result: Point3d[] = [];
  const append = (segment: CurveSegment3d, startS: number, start: CurveState3d, endS: number, end: CurveState3d, depth: number): void => {
    const middleS = (startS + endS) * 0.5;
    const middle = evaluateSegment(segment, middleS);
    const error = Math.hypot(
      middle.position[0] - (start.position[0] + end.position[0]) * 0.5,
      middle.position[1] - (start.position[1] + end.position[1]) * 0.5,
    );
    if (depth < 20 && (error > Math.max(maxChordErrorM, 1e-9) || endS - startS > Math.max(maxSegmentLengthM, 1e-6))) {
      append(segment, startS, start, middleS, middle, depth + 1);
      append(segment, middleS, middle, endS, end, depth + 1);
      return;
    }
    result.push(end.position);
  };
  for (const segment of geometry.segments) {
    const start = evaluateSegment(segment, 0);
    const previous = result[result.length - 1];
    if (!previous || Math.hypot(previous[0] - start.position[0], previous[1] - start.position[1], previous[2] - start.position[2]) > 1e-6) {
      result.push(start.position);
    }
    if (segment.lengthM > 0) append(segment, 0, start, segment.lengthM, evaluateSegment(segment, segment.lengthM), 0);
  }
  return result;
}

export function pathMidpoint(geometry: PathGeometry3d): Point3d {
  return evaluatePath(geometry, pathLength(geometry) * 0.5).position;
}

export function boundsOfPoints(points: Point3d[]): Bounds2d | undefined {
  if (points.length === 0) return undefined;
  const bounds = { minX: points[0][0], minY: points[0][1], maxX: points[0][0], maxY: points[0][1] };
  for (const point of points) {
    bounds.minX = Math.min(bounds.minX, point[0]); bounds.minY = Math.min(bounds.minY, point[1]);
    bounds.maxX = Math.max(bounds.maxX, point[0]); bounds.maxY = Math.max(bounds.maxY, point[1]);
  }
  return bounds;
}

function squaredDistanceToSegment(point: Point3d, first: Point3d, second: Point3d): number {
  const dx = second[0] - first[0];
  const dy = second[1] - first[1];
  if (dx === 0 && dy === 0) return (point[0] - first[0]) ** 2 + (point[1] - first[1]) ** 2;
  const ratio = Math.min(Math.max(((point[0] - first[0]) * dx + (point[1] - first[1]) * dy) / (dx * dx + dy * dy), 0), 1);
  const x = first[0] + dx * ratio;
  const y = first[1] + dy * ratio;
  return (point[0] - x) ** 2 + (point[1] - y) ** 2;
}

export function simplifyPolyline(points: Point3d[], toleranceM: number): Point3d[] {
  if (points.length <= 2 || toleranceM <= 0) return points.map((point) => [...point] as Point3d);
  const keep = new Uint8Array(points.length);
  keep[0] = 1;
  keep[points.length - 1] = 1;
  const stack: Array<[number, number]> = [[0, points.length - 1]];
  const threshold = toleranceM * toleranceM;
  while (stack.length > 0) {
    const [start, end] = stack.pop()!;
    let farthest = -1;
    let distance = threshold;
    for (let index = start + 1; index < end; index += 1) {
      const candidate = squaredDistanceToSegment(points[index], points[start], points[end]);
      if (candidate > distance) {
        distance = candidate;
        farthest = index;
      }
    }
    if (farthest >= 0) {
      keep[farthest] = 1;
      stack.push([start, farthest], [farthest, end]);
    }
  }
  return points.filter((_, index) => keep[index] === 1).map((point) => [...point] as Point3d);
}
